psvpfsparser/Utils.cpp

```diff
@@ -65,35 +65,28 @@ void getFileListNoPfs(boost::filesystem::path root_path, std::set<boost::filesys
          const boost::filesystem::path cp = (*i);
 
          //skip paths that are not included in files.db
+         //i.nopush(true) will skip recursion into directory
 
          //skip pfs directory
-         if(boost::starts_with(cp, (root_path / boost::filesystem::path("sce_pfs"))))
+         if(cp.filename() == boost::filesystem::path("sce_pfs")) {
+            i.no_push(true);
             continue;
+         }
 
          //skip packages
-         if(boost::starts_with(cp, (root_path / boost::filesystem::path("sce_sys") / boost::filesystem::path("package"))))
-            continue;
-
-         //skip pfs inside sce_pfs (for ADDCONT)
-         if(boost::ends_with(cp, boost::filesystem::path("sce_pfs")))
-            continue;
-
-         //skip pfs inside sce_pfs (for ADDCONT)
-         if(boost::ends_with(cp, boost::filesystem::path("sce_pfs") / boost::filesystem::path("files.db")))
-            continue;
-
-         //skip pfs inside sce_pfs (for ADDCONT)
-         if(boost::ends_with(cp, boost::filesystem::path("sce_pfs") / boost::filesystem::path("unicv.db")))
+         if(cp == (root_path / boost::filesystem::path("sce_sys/package"))) {
+            i.no_push(true);
             continue;
+         }
 
          //skip pfs inside sce_sys (for ADDCONT)
-         if(boost::ends_with(cp, boost::filesystem::path("sce_sys")))
-         {
-            if(cp != root_path / boost::filesystem::path("sce_sys"))
-               continue;
+         if(boost::ends_with(cp, boost::filesystem::path("sce_sys")) &&
+               cp != root_path / boost::filesystem::path("sce_sys") &&
+               boost::filesystem::exists(cp / boost::filesystem::path("keystone"))) {
```

Comment:
Can you explain why do we need to check keystone existence?

Reply:
I believe sce_sys directories which exist in the same directory as a sce_pfs directory will always contain a keystone file.

This particular check is for skipping sce_sys directories of the form `<TITLE ID>/<ADDCONT ID>/sce_sys` when decrypting an addcont root. If the keystone existence check is not there, then this would also skip for example `PCSG01136/savedata/sce_sys/param.sfo`.

Possibly a better approach that could be implemented in the future is to use files.db to determine which files are part of the pfs instead of filtering by names and paths.

Reply:
Yes. Keystone file is always inside sce_pfs directory and is always checked after decryption.
This check is done for both icv and unicv.

About usage of files.db. What you suggest is actually already implemented. However this `getFileListNoPfs` method is also required. Everything is a bit more complex. When parser finishes parsing files.db it validates the number of files.
Basically this helps us to verify that there are no extra files present in real filesystem which are not present in files.db. And the other way around as well. Parser intersects the list of real files vs list of files from files.db. However it excludes (masks) files that exist in files.db and not in real filesystem (nonexistent files can be specified in files.db). It also excludes (masks) files that exist in real filesystem but are expected not to be present in files.db. For example contents of sce_pfs directory. You can see this logic in the end of 'int FilesDbParser::parse()' method.